app/jobs/amend_backorder_job.rb

```diff
@@ -15,81 +16,119 @@ def perform(order)
+      solution = broker.best_offer(link)
+      line = orderer.find_or_build_order_line(backorder, solution.offer)
+      if variant.on_demand
+        adjust_stock(variant, solution, line)
+      else
+        aggregate_final_quantities(order_cycle, line, variant, solution)
+      end
+
+      line
+    end
+  end
 
-    order.lines.each do |line|
-      line.quantity = line.quantity.to_i
+  def cancel_stale_lines(unprocessed_lines, order, broker)
+    managed_variants = managed_linked_variants(order)
+    unprocessed_lines.each do |line|
+      wholesale_quantity = line.quantity.to_i
       wholesale_product_id = line.offer.offeredItem.semanticId
       transformation = broker.wholesale_to_retail(wholesale_product_id)
-      linked_variant = variants.linked_to(transformation.retail_product_id)
+      linked_variant = managed_variants.linked_to(transformation.retail_product_id)
 
-      # Assumption: If a transformation is present then we only sell the retail
-      # variant. If that can't be found, it was deleted and we'll ignore that
-      # for now.
-      next if linked_variant.nil?
+      if linked_variant.nil?
+        transformation.factor = 1
+        linked_variant = managed_variants.linked_to(wholesale_product_id)
+      end
 
-      # Find all line items for this order cycle
-      # Update quantity accordingly
-      if linked_variant.on_demand
-        release_superfluous_stock(line, linked_variant, transformation)
-      else
-        aggregate_final_quantities(order_cycle, line, linked_variant, transformation)
+      # Adjust stock level back, we're not going to order this one.
+      if linked_variant&.on_demand
+        retail_quantity = wholesale_quantity * transformation.factor
+        linked_variant.on_hand -= retail_quantity
       end
-    end
 
-    # Clean up empty lines:
-    order.lines.reject! { |line| line.quantity.zero? }
+      # We don't have any active orders for this
+      line.quantity = 0
+    end
   end
 
-  # We look at all linked variants.
-  def backorderable_items(order)
-    order.line_items.select do |item|
-      # TODO: scope variants to hub.
-      # We are only supporting producer stock at the moment.
-      item.variant.semantic_links.present?
+  def adjust_stock(variant, solution, line)
+    if variant.on_hand.negative?
+      needed_quantity = -1 * variant.on_hand # We need to replenish it.
+
+      # The number of wholesale packs we need to order to fulfill the
+      # needed quantity.
+      # For example, we order 2 packs of 12 cans if we need 15 cans.
+      wholesale_quantity = (needed_quantity.to_f / solution.factor).ceil
+
+      # The number of individual retail items we get with the wholesale order.
+      # For example, if we order 2 packs of 12 cans, we will get 24 cans
+      # and we'll account for that in our stock levels.
+      retail_quantity = wholesale_quantity * solution.factor
+
+      line.quantity = line.quantity.to_i + wholesale_quantity
```

Comment:
I'm curious why we can't increment quantity directly?

```suggestion
      line.quantity += wholesale_quantity
```

Reply:
looks like we can, in future commit.

Reply:
The Shopify API is providing the quantity as a string and we need to cast it to an integer first. But later I did that in a separate line.